File: hard/kotlin/abc127_d.kt
/**
 * 6
 * ABC127/D
 * 貪欲法
 */

fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val A = readLine()!!.split(" ").map(String::toLong).sorted().toMutableList()
  val BC = List(M) { readLine()!!.split(" ").map(String::toInt) }.sortedByDescending { it[1] }

  var i = 0
  BC.forEach {(b, c) ->
    var _b = b
    while (i < N && _b-- > 0 && A[i] < c) {
      A[i++] = c.toLong()
    }
  }
  println(A.sum())
}

File: hard/ruby/abc127_d.rb
# 6
# ABC127/D
# 貪欲法

N, M = gets.split.map(&:to_i)
A = gets.split.map(&:to_i).sort
BC = Array.new(M) { gets.split.map(&:to_i) }.sort_by { _1[1] }.reverse

i = 0
BC.each do |b, c|
  while i < N && b > 0 && A[i] < c
    A[i] = c
    b -= 1
    i += 1
  end
end

puts A.sum
